File: lib/bridge/payload.ts
/**
 * SyncPayload — shape posted by the Python bridge to `/api/bridge/sync`.
 * See docs/13-bridge.md for the reference spec.
 *
 * All validation is structural (presence + type). Value-level sanity checks
 * belong in the engines that consume this data (e.g. x_ratio ∈ [0,1] is
 * enforced at the DB CHECK constraint, not here).
 */
import type {
  CombinationFactor,
  LoadType,
  MemberType,
  SupportType,
} from '@/lib/supabase/types'

export type SyncNode = {
  node_id: number
  x_mm: number
  y_mm: number
  z_mm: number
  support_type: SupportType
}

export type SyncMember = {
  member_id: number
  start_node_id: number
  end_node_id: number
  section_name: string
  material_name?: string | null
  length_mm: number
  beta_angle_deg?: number
  member_type: MemberType
}

export type SyncSection = {
  section_name: string
  section_type: 'rectangular' | 'i_section' | 'circular'
  b_mm?: number | null
  h_mm?: number | null
  area_mm2?: number | null
  i_major_mm4?: number | null
  i_minor_mm4?: number | null
}

export type SyncMaterial = {
  name: string
  e_mpa: number
  density_kn_m3: number
  fc_mpa?: number | null
  fy_mpa?: number | null
}

export type SyncLoadCase = {
  case_number: number
  title: string
  load_type: LoadType
}

export type SyncCombination = {
  combo_number: number
  title: string
  factors: CombinationFactor[]
  source?: 'imported' | 'app_generated'
}

export type SyncDiagramPoint = {
  member_id: number
  combo_number: number
  x_ratio: number
  x_mm: number
  mz_knm: number
  vy_kn: number
  n_kn: number
}

export type SyncEnvelope = {
  member_id: number
  mpos_max_knm: number
  mpos_combo?: number | null
  mneg_max_knm: number
  mneg_combo?: number | null
  vu_max_kn: number
  vu_combo?: number | null
  nu_tension_max_kn?: number
  nu_compression_max_kn?: number
}

export type SyncReaction = {
  node_id: number
  combo_number: number
  rx_kn: number
  ry_kn: number
  rz_kn: number
  mx_knm: number
  my_knm: number
  mz_knm: number
}

export type SyncPayload = {
  project_id: string
  file_name: string
  file_hash: string
  nodes: SyncNode[]
  members: SyncMember[]
  sections: SyncSection[]
  materials?: SyncMaterial[]
  load_cases: SyncLoadCase[]
  combinations: SyncCombination[]
  diagram_points: SyncDiagramPoint[]
  envelope: SyncEnvelope[]
  reactions: SyncReaction[]
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export class PayloadError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'PayloadError'
  }
}

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new PayloadError(message)
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

function isArray<T>(v: unknown): v is T[] {
  return Array.isArray(v)
}

function isNum(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v)
}

function isStr(v: unknown): v is string {
  return typeof v === 'string'
}

/**
 * Shallow structural validation. Returns the payload typed — does NOT coerce.
 * Throws PayloadError with a specific message on the first bad field.
 */
export function parseSyncPayload(body: unknown): SyncPayload {
  assert(isObject(body), 'payload must be an object')

  assert(isStr(body.project_id), 'project_id must be a string')
  assert(isStr(body.file_name), 'file_name must be a string')
  assert(isStr(body.file_hash), 'file_hash must be a string')

  const arrayFields = [
    'nodes',
    'members',
    'sections',
    'load_cases',
    'combinations',
    'diagram_points',
    'envelope',
    'reactions',
  ] as const
  for (const key of arrayFields) {
    assert(isArray(body[key]), `${key} must be an array`)
  }
  if (body.materials !== undefined) {
    assert(isArray(body.materials), 'materials must be an array when provided')
  }

  // Spot-check a representative item per array — cheap, catches the common
  // shape errors from an out-of-date bridge build without O(n) work.
  const nodes = body.nodes as unknown[]
  if (nodes.length > 0) {
    const n = nodes[0] as Record<string, unknown>
    assert(isNum(n.node_id) && isNum(n.x_mm) && isNum(n.y_mm) && isNum(n.z_mm),
      'nodes[0] has missing/invalid coordinates')
  }

  const diagrams = body.diagram_points as unknown[]
  if (diagrams.length > 0) {
    const d = diagrams[0] as Record<string, unknown>
    assert(
      isNum(d.member_id) && isNum(d.combo_number) &&
      isNum(d.x_ratio) && isNum(d.x_mm) &&
      isNum(d.mz_knm) && isNum(d.vy_kn),
      'diagram_points[0] has missing/invalid fields',
    )
  }

  return body as unknown as SyncPayload
}

File: staad-bridge/payload.py
"""Pydantic models mirroring the SyncPayload contract in docs/13-bridge.md.

Kept as plain models (not inherited from a common base) so the JSON
serialisation matches the TS shape exactly. Field names stay lowercase
snake_case because the Supabase columns do — see docs/12-conventions.md.
"""
from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel


LoadType = Literal[
    "dead", "live", "roof_live", "wind_x", "wind_z",
    "seismic_x", "seismic_z", "other",
]
MemberType = Literal["beam", "column", "brace", "other"]
SupportType = Literal["fixed", "pinned", "roller_x", "roller_z"]
SectionType = Literal["rectangular", "i_section", "circular"]


class SyncNode(BaseModel):
    node_id: int
    x_mm: float
    y_mm: float
    z_mm: float
    support_type: Optional[SupportType] = None


class SyncMember(BaseModel):
    member_id: int
    start_node_id: int
    end_node_id: int
    section_name: str
    material_name: Optional[str] = None
    length_mm: float
    beta_angle_deg: float = 0.0
    member_type: MemberType


class SyncSection(BaseModel):
    section_name: str
    section_type: SectionType
    b_mm: Optional[float] = None
    h_mm: Optional[float] = None
    area_mm2: Optional[float] = None
    i_major_mm4: Optional[float] = None
    i_minor_mm4: Optional[float] = None


class SyncMaterial(BaseModel):
    name: str
    e_mpa: float
    density_kn_m3: float
    fc_mpa: Optional[float] = None
    fy_mpa: Optional[float] = None


class SyncLoadCase(BaseModel):
    case_number: int
    title: str
    load_type: LoadType


class CombinationFactor(BaseModel):
    case_number: int
    load_type: LoadType
    factor: float


class SyncCombination(BaseModel):
    combo_number: int
    title: str
    factors: List[CombinationFactor]
    source: Literal["imported", "app_generated"] = "imported"


class SyncDiagramPoint(BaseModel):
    member_id: int
    combo_number: int
    x_ratio: float  # 0.0..1.0, 11 samples per member per combo
    x_mm: float
    mz_knm: float
    vy_kn: float
    n_kn: float = 0.0
    # Minor-axis moment + shear (required for biaxial column check).
    my_knm: float = 0.0
    vz_kn: float = 0.0


class SyncEnvelope(BaseModel):
    member_id: int
    mpos_max_knm: float
    mpos_combo: Optional[int] = None
    mneg_max_knm: float
    mneg_combo: Optional[int] = None
    vu_max_kn: float
    vu_combo: Optional[int] = None
    nu_tension_max_kn: float = 0.0
    nu_compression_max_kn: float = 0.0
    # Minor-axis peaks for the biaxial column path.
    mpos_max_minor_knm: float = 0.0
    mpos_combo_minor: Optional[int] = None
    mneg_max_minor_knm: float = 0.0
    mneg_combo_minor: Optional[int] = None


class SyncReaction(BaseModel):
    node_id: int
    combo_number: int
    rx_kn: float = 0.0
    ry_kn: float = 0.0
    rz_kn: float = 0.0
    mx_knm: float = 0.0
    my_knm: float = 0.0
    mz_knm: float = 0.0


class SyncPayload(BaseModel):
    """Top-level payload POSTed to /api/bridge/sync."""

    project_id: str
    file_name: str
    file_hash: str
    unit_system: str = "unknown"
    nodes: List[SyncNode]
    members: List[SyncMember]
    sections: List[SyncSection]
    materials: List[SyncMaterial] = []
    load_cases: List[SyncLoadCase]
    combinations: List[SyncCombination]
    diagram_points: List[SyncDiagramPoint]
    envelope: List[SyncEnvelope]
    reactions: List[SyncReaction]


# --- Inbound shapes (app → bridge) ------------------------------------------

class ResyncBody(BaseModel):
    project_id: str


class PushCombinationsBody(BaseModel):
    project_id: str
    combinations: List[SyncCombination]
